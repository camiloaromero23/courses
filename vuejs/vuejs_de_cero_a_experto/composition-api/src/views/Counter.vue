<template>
  <h1>Counter view</h1>
  <h3>{{ counter }}</h3>
  <button @click="decrease">-</button>
  <button @click="increase">+</button>
</template>

<script>
import useCounter from "@/composables/useCounter";

export default {
  name: "Counter",
  props: {},
  emits: [],
  setup() {
    const { counter, increase, decrease } = useCounter(30);

    return {
      counter,
      increase,
      decrease,
    };
  },
};
</script>

<style></style>
