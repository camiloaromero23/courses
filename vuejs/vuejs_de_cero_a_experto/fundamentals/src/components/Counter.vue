<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>
  <p data-testid="counter">{{ counter }}</p>
  <div>
    <button @click="increaseCounter">+</button>
    <button @click="decreaseCounter">-</button>
    <button @click="resetCounter">Clear</button>
  </div>
</template>

<script>
export default {
  name: 'Counter',
  props: {
    title: String,
    start: {
      type: Number,
      default: 0,
      /* required: true, */
      validator(value) {
        return value >= 0;
      },
    },
  },
  data() {
    return {
      counter: this.start,
    };
  },
  methods: {
    increaseCounter() {
      this.counter += 1;
    },
    decreaseCounter() {
      this.counter -= 1;
    },
    resetCounter() {
      this.counter = 0;
    },
    /* getSquareValue() { */
    /*   return this.counter * this.counter; */
    /* }, */
  },
  computed: {
    squareCounter() {
      return this.counter * this.counter;
    },
    customTitle() {
      return this.title || 'Counter';
    },
  },
};
</script>

<style>
button {
  background-color: #64bb87;
  border-radius: 5px;
  border: 1px solid white;
  color: white;
  cursor: pointer;
  margin: 0 5px;
  padding: 5px 15px;
  transition: 0.3s ease-in-out;
}

button:hover {
  background-color: #5aa67b;
  transition: 0.3s;
}
</style>
