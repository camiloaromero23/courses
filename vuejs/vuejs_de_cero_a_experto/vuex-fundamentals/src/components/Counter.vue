<template>
  <h1>Counter - Vuex</h1>
  <h2>Direct access: {{ $store.state.counter.count }}</h2>
  <h2>Computed: {{ computedCount }}</h2>

  <button @click="increase">+1</button>
  <button @click="increaseBy5">+5</button>
  <button @click="randomIncrease" :disabled="loading">Random</button>

  <h1>mapState counter: {{ count }}</h1>
  <h1>mapState mutation: {{ lastMutation }}</h1>

  <h2>Direct getter: {{ $store.getters["counter/squaredCount"] }}</h2>
  <h2>mapGetters squaredCount: {{ squaredCount }}</h2>
</template>

<script>
import { mapState, mapActions, mapGetters } from "vuex";
export default {
  computed: {
    computedCount() {
      return this.$store.state.counter.count;
    },
    /* ...mapState('counter',["count", "lastMutation", "loading"]) */
    ...mapState("counter", {
      count: state => state.count,
      lastMutation: state => state.lastMutation,
      loading: state => state.loading
      /* loading: "loading", */
      /* lastMutation: "lastMutation" */
    }),
    ...mapGetters("counter", {
      squaredCount: "squaredCount"
    })
  },
  methods: {
    increase() {
      this.$store.commit("counter/increase");
    },
    increaseBy5() {
      this.$store.commit("counter/increaseBy", 5);
      /* this.randomIncrease(); // Can also dispatch an action this way */
    },
    /* ...mapActions("counter", ["increaseByRandomInt"]) */
    ...mapActions("counter", {
      randomIncrease: "increaseByRandomInt"
    })
  }
};
</script>
