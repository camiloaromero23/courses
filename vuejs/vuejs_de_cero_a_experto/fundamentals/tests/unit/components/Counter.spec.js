import { shallowMount } from '@vue/test-utils';
import Counter from '@/components/Counter';

describe('Counter Component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Counter);
  });

  test('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should have "Counter" as default value of h2', () => {
    const h2Tag = wrapper.find('h2');

    expect(h2Tag.exists()).toBeTruthy();

    expect(h2Tag.text()).toBe('Counter');
  });

  test('should have 0 as default value inside the p', () => {
    const value = wrapper.find('[data-testid="counter"]').text();

    expect(value).toContain('0');
  });

  test('should increase & decrease by 1 the counter', async () => {
    const [increaseButton, decreaseButton] = wrapper.findAll('button');

    await increaseButton.trigger('click');
    await increaseButton.trigger('click');
    await increaseButton.trigger('click');
    await decreaseButton.trigger('click');
    await decreaseButton.trigger('click');

    const value = wrapper.find('[data-testid="counter"]').text();

    expect(value).toBe('1');
  });

  test('should establish default value', () => {
    const { start } = wrapper.props();

    const value = wrapper.find('[data-testid="counter"]').text();

    expect(+value).toBe(start);
  });

  test('should show title prop', () => {
    const title = 'Test Counter';
    const wrapper = shallowMount(Counter, {
      props: {
        title,
      },
    });

    expect(wrapper.find('h2').text()).toBe(title);
  });
});
