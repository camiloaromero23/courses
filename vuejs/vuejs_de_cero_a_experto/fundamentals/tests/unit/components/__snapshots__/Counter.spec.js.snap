// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component should match the snapshot 1`] = `
<h2>Counter</h2>
<p>0 <sup>2</sup> = 0</p>
<p data-testid="counter">0</p>
<div><button>+</button><button>-</button><button>Clear</button></div>
`;
